From dd351e580616c88c5b78a20b83b36ac0fa27f411 Mon Sep 17 00:00:00 2001
From: Abhishek Chauhan <quic_abchauha@quicinc.com>
Date: Sat, 18 Jan 2025 15:12:40 -0800
Subject: [PATCH 04/16] PENDING: net: pcs: Add 5G baseR changes in PCS driver

Add 5G baseR changes in PCS driver and disable clause 37
autonegotiation to support Fixed link configuration.

Change-Id: I7d69b3d891ac621b1cb04e586d20cd20a391d60f
Signed-off-by: Abhishek Chauhan <quic_abchauha@quicinc.com>
(cherry picked from commit fcdee4104c9a14e8df6f7675684e3e1f8c9a61f3)
---
 drivers/net/pcs/pcs-xpcs-qcom.c | 41 ++++++++++++++++++++++++++++++++-
 1 file changed, 40 insertions(+), 1 deletion(-)

diff --git a/drivers/net/pcs/pcs-xpcs-qcom.c b/drivers/net/pcs/pcs-xpcs-qcom.c
index c625a8247ecb..37caf6256797 100644
--- a/drivers/net/pcs/pcs-xpcs-qcom.c
+++ b/drivers/net/pcs/pcs-xpcs-qcom.c
@@ -54,6 +54,21 @@ static const int xpcs_10gbaser_features[] = {
 	__ETHTOOL_LINK_MODE_MASK_NBITS,
 };
 
+static const int xpcs_usx5g_features[] = {
+	ETHTOOL_LINK_MODE_Pause_BIT,
+	ETHTOOL_LINK_MODE_Asym_Pause_BIT,
+	ETHTOOL_LINK_MODE_Autoneg_BIT,
+	ETHTOOL_LINK_MODE_10baseT_Half_BIT,
+	ETHTOOL_LINK_MODE_10baseT_Full_BIT,
+	ETHTOOL_LINK_MODE_100baseT_Half_BIT,
+	ETHTOOL_LINK_MODE_100baseT_Full_BIT,
+	ETHTOOL_LINK_MODE_1000baseT_Half_BIT,
+	ETHTOOL_LINK_MODE_1000baseT_Full_BIT,
+	ETHTOOL_LINK_MODE_2500baseT_Full_BIT,
+	ETHTOOL_LINK_MODE_5000baseT_Full_BIT,
+	__ETHTOOL_LINK_MODE_MASK_NBITS,
+};
+
 static const phy_interface_t xpcs_usxgmii_interfaces[] = {
 	PHY_INTERFACE_MODE_USXGMII,
 };
@@ -62,9 +77,14 @@ static const phy_interface_t xpcs_10gbaser_interfaces[] = {
 	PHY_INTERFACE_MODE_10GBASER,
 };
 
+static const phy_interface_t xpcs_usx5g_interfaces[] = {
+	PHY_INTERFACE_MODE_5GBASER,
+};
+
 enum {
 	DW_XPCS_USXGMII,
 	DW_XPCS_10GBASER,
+	DW_XPCS_USX5G,
 	DW_XPCS_INTERFACE_MAX,
 };
 
@@ -485,7 +505,9 @@ static int qcom_xpcs_select_mode(struct dw_xpcs_qcom *qxpcs, phy_interface_t int
 
 	g_interface = interface;
 
-	if (interface == PHY_INTERFACE_MODE_USXGMII || interface == PHY_INTERFACE_MODE_10GBASER) {
+	if (interface == PHY_INTERFACE_MODE_USXGMII ||
+	    interface == PHY_INTERFACE_MODE_10GBASER ||
+	    interface == PHY_INTERFACE_MODE_5GBASER) {
 		ret = qcom_xpcs_read(qxpcs, DW_SR_MII_PCS_CTRL2);
 		if (ret < 0)
 			goto out;
@@ -515,6 +537,16 @@ static int qcom_xpcs_select_mode(struct dw_xpcs_qcom *qxpcs, phy_interface_t int
 		if (interface == PHY_INTERFACE_MODE_USXGMII)
 			qcom_xpcs_write(qxpcs, DW_VR_MII_PCS_DIG_CTRL1, ret | DW_USXGMII_EN);
 
+		ret = qcom_xpcs_read(qxpcs, DW_VR_MII_PCS_KR_CTRL);
+		if (ret < 0)
+			goto out;
+
+		ret &= ~USXG_MODE_SEL;
+		if (interface == PHY_INTERFACE_MODE_5GBASER)
+			ret |= USXGMII_5G;
+
+		qcom_xpcs_write(qxpcs, DW_VR_MII_PCS_KR_CTRL, ret);
+
 		return 0;
 	}
 
@@ -542,6 +574,7 @@ void qcom_xpcs_link_up(struct phylink_pcs *pcs, unsigned int mode,
 
 	switch (interface) {
 	case PHY_INTERFACE_MODE_USXGMII:
+	case PHY_INTERFACE_MODE_5GBASER:
 		qcom_xpcs_link_up_usxgmii(qxpcs, speed);
 		break;
 	default:
@@ -587,6 +620,12 @@ static const struct xpcs_compat synopsys_xpcs_compat[DW_XPCS_INTERFACE_MAX] = {
 		.num_interfaces = ARRAY_SIZE(xpcs_10gbaser_interfaces),
 		.an_mode = DW_10GBASER,
 	},
+	[DW_XPCS_USX5G] = {
+		.supported = xpcs_usx5g_features,
+		.interface = xpcs_usx5g_interfaces,
+		.num_interfaces = ARRAY_SIZE(xpcs_usx5g_interfaces),
+		.an_mode = DW_10GBASER,
+	},
 };
 
 static const struct xpcs_id xpcs_id_list[] = {
